### PR TITLE
assignemt 4

### PR DESCRIPTION
**Pull request by anaritarc**

### CODE REVIEW

File: life_expectancy/load_save_data.py [hunk 1]
```diff
@@ -1,18 +1,36 @@
...
-    """Save data in a csv"""
+class LoadJsonStrategy(LoadStrategy):
+    """Context Strategy Class to Load Json files"""
+    def load_data(self):
+        name_file = DATA_DIR / "eurostat_life_expect.zip"
+        with zipfile.ZipFile(name_file, 'r') as zip_ref:
+            json_filename = zip_ref.namelist()[0]
+            zip_ref.extract(json_filename)
+            with open(json_filename, 'r', encoding='utf-8') as json_file:
+                data = json.load(json_file)
```
Comment (atneves) [inline life_expectancy/load_save_data.py]: I think you can use pd.read_json with compression directly

File: life_expectancy/pipeline.py [hunk 2]
```diff
@@ -1,17 +1,80 @@
...
+    Belarus = 'BY'
+    United_Kingdom = 'UK'
+    Moldova = 'MD'
+    Russia = 'RU'
+
+
+def main(region='pt', json_csv='csv') -> None:
+    strategies = {
+        'json': LoadJsonStrategy(),
+        'csv': LoadCsvStrategy()
```
Comment (atneves) [inline life_expectancy/pipeline.py]: Maybe you can do a dictionary like this : 
strategies = {
        'json': LoadJsonStrategy,
        'csv': LoadCsvStrategy
}
and only instatiate the classes when you need it

File: life_expectancy/tests/test_load_save_data.py [hunk 3]
```diff
@@ -1,15 +1,23 @@
...
 
 
 @patch("life_expectancy.load_save_data.pd.read_table")
-def test_load_data(mock_read_csv, eu_life_expectancy_raw) -> None:
+def test_load_json(mock_read_json, eurostat_life_expect) -> None:
+    """test load_data is a panda dataframe"""
+    mock_read_json.return_value = eurostat_life_expect
+    strategy = LoadJsonStrategy()
+    result = strategy.load_data()
+    assert isinstance(result, pd.DataFrame)
```
Comment (atneves) [inline life_expectancy/tests/test_load_save_data.py]: you can assert it is not null too

File: life_expectancy/tests/test_load_save_data.py [hunk 4]
```diff
@@ -1,15 +1,23 @@
...
+
+@patch("life_expectancy.load_save_data.pd.read_csv")
+def test_load_csv(mock_read_csv, eu_life_expectancy_raw) -> None:
     """test load_data is a panda dataframe"""
     mock_read_csv.return_value = eu_life_expectancy_raw
-    result = load_data()
+    result = LoadCsvStrategy().load_data()
```
Comment (atneves) [inline life_expectancy/tests/test_load_save_data.py]: I think it is possible to join this 2 load tests in one using @pytest.mark.parametrize

**Review by atneves — COMMENTED** [review]
Good work :)

**Review by atneves — COMMENTED** [review]
Good work :)

File: life_expectancy/tests/test_pipeline.py [hunk 5]
```diff
@@ -1,24 +1,31 @@
...
     @patch('life_expectancy.pipeline.save_data')
-    @patch('argparse.ArgumentParser.parse_args')
-    def test_main(self, mock_parse_args, mock_save_data, mock_clean_data, mock_load_data):
-        mock_parse_args.return_value = Namespace(region='pt')
-        mock_load_data.return_value = 'raw_data'
-        mock_clean_data.return_value = 'cleaned_data'
-        main()
-        mock_load_data.assert_called_once()
+    def test_main(self, mock_save_data, mock_clean_data, mock_load_strategy):
+        mock_load_data = MagicMock()
```
Comment (MrCordeiro) [inline life_expectancy/tests/test_pipeline.py:12]: **Suggestion** Try running without this

File: life_expectancy/tests/test_pipeline.py [hunk 6]
```diff
@@ -1,24 +1,31 @@
...
+        mock_load_data.load_data.return_value = 'raw_data'
+        mock_load_strategy.return_value = mock_load_data
+        mock_cleaned_data = MagicMock()
+        mock_clean_data.return_value = mock_cleaned_data
+        mock_save_data.return_value = None
+
+        args = Namespace(region='pt', json_csv='json')
+        main(args.region, args.json_csv)
```
Comment (MrCordeiro) [inline life_expectancy/tests/test_pipeline.py]:
```suggestion
        main('pt', 'json')
```

File: life_expectancy/tests/conftest.py [hunk 7]
```diff
@@ -20,6 +20,17 @@ def eu_life_expectancy_raw() -> pd.DataFrame:
...
+@pytest.fixture(scope="session")
+def eurostat_life_expect() -> pd.DataFrame:
+    """Fixture to load the expected output of the cleaning script"""
+    zip_file_path = FIXTURES_DIR / "eurostat_life_expect.zip"
+    with zipfile.ZipFile(zip_file_path) as zip_ref:
+        json_filename = zip_ref.namelist()[0]
+        zip_ref.extract(json_filename)
+        with zip_ref.open(json_filename) as json_file:
+            json_data = json_file.read().decode('utf-8')
+            return pd.read_json(json_data)
```
Comment (MrCordeiro) [inline life_expectancy/tests/conftest.py]:
```suggestion
     return pd.read_json(zip_file_path, compression="infer")
```

File: life_expectancy/tests/conftest.py [hunk 8]
```diff
@@ -1,7 +1,7 @@
 """Pytest configuration file"""
+import zipfile
```
Comment (MrCordeiro) [inline life_expectancy/tests/conftest.py]:
```suggestion
```

File: life_expectancy/pipeline.py [hunk 9]
```diff
@@ -1,17 +1,80 @@
...
+    Ukraine = 'UA'
+    Belarus = 'BY'
+    United_Kingdom = 'UK'
+    Moldova = 'MD'
+    Russia = 'RU'
+
+
+def main(region='pt', json_csv='csv') -> None:
+    strategies = {
+        'json': LoadJsonStrategy(),
```
Comment (MrCordeiro) [inline life_expectancy/pipeline.py]: `LoadZippedJsonStrategy`

If you receive a Path, you can use Path.suffix to extract your strategy key

File: life_expectancy/pipeline.py [hunk 10]
```diff
@@ -1,17 +1,80 @@
...
+    Serbia = 'RS'
+    Armenia = 'AM'
+    Azerbaijan = 'AZ'
+    Georgia = 'GE'
+    Turkey = 'TR'
+    Ukraine = 'UA'
+    Belarus = 'BY'
+    United_Kingdom = 'UK'
+    Moldova = 'MD'
+    Russia = 'RU'
```
Comment (MrCordeiro) [inline life_expectancy/pipeline.py:54]: Missing method to return all contry values as list of strings

File: life_expectancy/pipeline.py [hunk 11]
```diff
@@ -1,17 +1,80 @@
...
     save_data(data_by_region, region)
 
 
 if __name__ == '__main__': # pragma: no cover
     parser = argparse.ArgumentParser()
-    parser.add_argument("-r","--region", help="Select data region", type=str, default ='pt')
+    parser.add_argument(
+        "-r","--region", 
+        help="Select data region",
+        type=Country, choices=list(Country),
```
Comment (MrCordeiro) [inline life_expectancy/pipeline.py:75]: Validate if list(Country) return a list of string

File: life_expectancy/load_save_data.py [hunk 12]
```diff
@@ -1,18 +1,36 @@
...
+        return pd.DataFrame(data)
+
+class LoadCsvStrategy(LoadStrategy):
+    """Strategy Class to Load csv files"""
+    def load_data(self):
+        name_file = DATA_DIR / "eu_life_expectancy_raw.tsv"
+        return pd.read_csv(name_file, sep='\t')
+
+def save_data(df, region) -> None:
```
Comment (MrCordeiro) [inline life_expectancy/load_save_data.py]: Missing type hints

File: life_expectancy/pipeline.py [hunk 13]
```diff
@@ -1,17 +1,80 @@
...
+    Georgia = 'GE'
+    Turkey = 'TR'
+    Ukraine = 'UA'
+    Belarus = 'BY'
+    United_Kingdom = 'UK'
+    Moldova = 'MD'
+    Russia = 'RU'
+
+
+def main(region='pt', json_csv='csv') -> None:
```
Comment (MrCordeiro) [inline life_expectancy/pipeline.py]: Should main be receiving a string or a Region?

**Review by MrCordeiro — CHANGES_REQUESTED** [review]

**Pull request closed without merging**